sources/packages/backend/apps/api/src/services/application/application.service.ts

```diff
@@ -921,39 +937,71 @@ export class ApplicationService extends RecordDataModelService<Application> {
         .getMany()
     );
   }
+
   /**
-   * Update Student Application status.
+   * Update Student Application status to cancelled.
    * Only allows the update on applications that are not in a final status.
    * The final statuses of an application are Completed, Overwritten and Cancelled.
    * @param applicationId application id.
-   * @param applicationStatus application status that need to be updated.
+   * @param studentId student id for authorization purposes.
    * @param auditUserId user who is making the changes.
-   * @returns student Application UpdateResult.
+   * @returns student application update result.
    */
-  async updateApplicationStatus(
+  async cancelStudentApplication(
     applicationId: number,
-    applicationStatus: ApplicationStatus,
+    studentId: number,
     auditUserId: number,
-  ): Promise<UpdateResult> {
-    const now = new Date();
-    return this.repo.update(
-      {
-        id: applicationId,
-        applicationStatus: Not(
-          In([
-            ApplicationStatus.completed,
-            ApplicationStatus.overwritten,
-            ApplicationStatus.cancelled,
-          ]),
-        ),
-      },
-      {
-        applicationStatus: applicationStatus,
-        applicationStatusUpdatedOn: now,
-        modifier: { id: auditUserId } as User,
-        updatedAt: now,
-      },
-    );
+  ): Promise<Application> {
+    return this.dataSource.transaction(async (transactionEntityManager) => {
+      const applicationRepo =
+        transactionEntityManager.getRepository(Application);
+      const application = await applicationRepo.findOne({
+        select: {
+          id: true,
+          currentAssessment: {
+            id: true,
+            assessmentWorkflowId: true,
+          },
+        },
+        relations: {
+          currentAssessment: true,
+        },
+        where: {
+          id: applicationId,
+          student: {
+            id: studentId,
+          },
+          applicationStatus: Not(
+            In([
+              ApplicationStatus.completed,
+              ApplicationStatus.overwritten,
+              ApplicationStatus.cancelled,
+            ]),
+          ),
+        },
+      });
+      if (!application) {
+        throw new CustomNamedError(
+          "Application not found or it is not in the correct state to be cancelled.",
+          APPLICATION_NOT_FOUND,
+        );
+      }
+      // Updates the application status to cancelled.
+      const now = new Date();
+      application.applicationStatus = ApplicationStatus.cancelled;
+      application.applicationStatusUpdatedOn = now;
+      application.modifier = { id: auditUserId } as User;
+      application.updatedAt = now;
+      await applicationRepo.save(application);
+      // Delete workflow and rollback overawards if the workflow started.
+      // Workflow doest not exists for draft or submitted application, for instance.
+      if (application.currentAssessment?.assessmentWorkflowId) {
```

Comment:
I have the same view here as we had for start application assessment. What if the transaction took a while to commit and before that queue started processing an assessment who's application is still not cancelled to rollback overawards. Or after execution of line 999, what happens if database goes down.?

Reply:
I see the point but there are a few considerations.
1 - This is different from the student submission because if the DB fails or is delayed the queue can still be processed with success. So the "saving to the DB" is not critical to keep the workflow cancelation from happening. In the Student application, if the application is not saved to submitted it will fail right at the start.
2 - The workflow cancellations are safe to be executed even in parallel, in the same way, as overawards cancellation. 
3 - In this way, if DB fails or "adds to the queue" fails, the student can still retry the operation.